crates/kitsune_p2p/bootstrap/src/sync.rs

```diff
@@ -0,0 +1,61 @@
+use std::sync::Arc;
+
+use super::*;
+use parking_lot::Mutex;
+use tokio::sync::Barrier;
+use warp::Filter;
+
+#[derive(Clone)]
+pub struct NodeSync {
+    waiting: Arc<Mutex<Option<Arc<Barrier>>>>,
+}
+
+impl NodeSync {
+    pub(crate) fn new() -> Self {
+        Self {
+            waiting: Arc::new(Mutex::new(None)),
+        }
+    }
+}
+
+pub(crate) fn sync(
+    sync: NodeSync,
+) -> impl Filter<Extract = impl warp::Reply, Error = warp::Rejection> + Clone {
+    warp::post()
+        .and(warp::header::exact("content-type", "application/octet"))
+        .and(warp::header::exact("X-Op", "sync"))
+        .and(warp::body::content_length_limit(SIZE_LIMIT))
+        .and(warp::body::bytes())
+        .and(with_sync(sync))
+        .and_then(waiting)
+}
+
+async fn waiting(query: Bytes, sync: NodeSync) -> Result<impl warp::Reply, warp::Rejection> {
+    let query: u64 = rmp_decode(&mut AsRef::<[u8]>::as_ref(&query)).map_err(|_| warp::reject())?;
+    let wait = {
+        let mut b = sync.waiting.lock();
+        match b.as_mut() {
+            Some(b) => b.clone(),
```

Comment:
throw an error if the incoming query count doesn't match the barrier's query count?